projects/igniteui-angular/src/lib/tabs/tabs.directive.ts

```diff
@@ -0,0 +1,318 @@
+import { IgxTabItemDirective } from './tab-item.directive';
+import { IgxTabContentBase, IgxTabsBase } from './tabs.base';
+
+export interface ITabsBaseEventArgs extends IBaseEventArgs {
+    readonly owner: IgxTabsDirective;
+}
+
+export interface ITabsSelectedIndexChangingEventArgs extends ITabsBaseEventArgs {
+    cancel: boolean;
+    readonly oldIndex: number;
+    newIndex: number;
+}
+
+export interface ITabsSelectedItemChangeEventArgs extends ITabsBaseEventArgs {
+    readonly oldItem: IgxTabItemDirective;
+    readonly newItem: IgxTabItemDirective;
+}
+
+@Directive()
+export abstract class IgxTabsDirective extends IgxCarouselComponentBase implements IgxTabsBase, AfterViewInit, OnDestroy {
+
+    /** @hidden */
+    @HostBinding('attr.role')
+    public role = 'tabs';
+
+    /**
+     * An @Input property that gets/sets the index of the selected item.
+     * Default value is 0 if contents are defined otherwise defaults to -1.
+     */
+    @Input()
+    public get selectedIndex(): number {
+        return this._selectedIndex;
+    }
+
+    public set selectedIndex(value: number) {
+        if (this._selectedIndex !== value) {
+            let newIndex = value;
+            const oldIndex = this._selectedIndex;
+            const args: ITabsSelectedIndexChangingEventArgs = {
+                owner: this,
+                cancel: false,
+                oldIndex,
+                newIndex
+            };
+            this.selectedIndexChanging.emit(args);
+
+            if (!args.cancel) {
+                newIndex = args.newIndex;
+                this._selectedIndex = newIndex;
+                this.selectedIndexChange.emit(this._selectedIndex);
+            }
+
+            this.updateSelectedTabs(oldIndex);
+        }
+    }
+
+    /**
+     * Enables/disables the transition animation of the contents.
+     */
+    @Input()
+    public get disableAnimation() {
+        return this._disableAnimation;
+    }
+
+    public set disableAnimation(value: boolean) {
+        this._disableAnimation = value;
+    }
+
+    /**
+     * Output to enable support for two-way binding on [(selectedIndex)]
+     */
+    @Output()
+    public selectedIndexChange = new EventEmitter<number>();
+
+    /**
+     * Emitted when the selected index is about to change.
+     */
+    @Output()
+    public selectedIndexChanging = new EventEmitter<ITabsSelectedIndexChangingEventArgs>();
+
+    /**
+     * Emitted when the selected item is changed.
+     */
+    @Output()
+    public selectedItemChange = new EventEmitter<ITabsSelectedItemChangeEventArgs>();
+
+    /**
+     * Returns the items.
+     */
+    @ContentChildren(IgxTabItemDirective)
+    public items: QueryList<IgxTabItemDirective>;
+
+    /**
+     * Gets the selected item.
+     */
+    public get selectedItem(): IgxTabItemDirective {
+        return this.items && this.selectedIndex >= 0 && this.selectedIndex < this.items.length ?
+            this.items.toArray()[this.selectedIndex] : null;

```

Comment:
```suggestion
            this.items.get(this.selectedIndex) : null;
```
Since 11.1, Angular supports `get` method on QueryList. No need to create a temporary array.

Reply:
Good to know :)